### ISSUE TITLE
Linktime dead code elimination only works on macOS

### ISSUE THREAD

**Issue by CaviarChen** [pschachte/wybe#60]
We added the support of linktime dead code elimination in #4 when wybe only works on macOS.
It turns out that the `-dead_strip` only exists in the system linker on macOS so it doesn't work on ubuntu.

I tried `-Wl,--gc-sections` and it doesn't work, I believe it requires that every function in its own ELF section.

**Comment by CaviarChen** [comment]
I took a further look into this issue and noticed the following things:

In terms of multiple specialization, we don't really need link-time dead code elimination.
Currently, multiple specialization only generates versions that we need. The missing specializations of a proc will be added into the module during the top-down pass, and we accept a module that has more specialization versions than we need (a superset). We can remove those during the top-down pass if we really don't want them and don't want to use link-time dead code elimination. That's super easy to do in the current top-down pass.

In terms of making the final executable smaller, the main factor is the lpvm section. The final executable contains a concatenation of all lpvm section of all object files it linked. This section usually takes 90% of the size of the executable and it is totally useless.


Back to the link-time dead code elimination.

The system linker on macOS provides an arg `-dead_strip` which can remove unused code and sections (such as the lpvm section) without any requirements.

On Linux, there isn't an equivalent arg. There is `-Wl,--gc-sections` which can remove unused code. However, it can only remove code by section, so we have to put procedures into different sections instead of a single section. For procedure foo, it should be in `.text.foo` instead of `.text`. This may make the object code a little bit bigger since there are more sections now. I tested this and it's doable.
Another problem is that `-Wl,--gc-sections` can't remove the lpvm section. I don't know why though.
To remove the lpvm section on Linux, I found the following two approaches:
1. Use cli `objcopy` to remove that section after we got the executable. It's simple but the problem is that the linker still needs to gather all lpvm sections and write to the executable file first.
2. Use a custom version of the linker script to discard this section, so the linker can skip it directly. It's nice but I am not sure what is the best practice of using a custom linker script. It seems a bit complicated.

Overall, I am not sure what approaches should we use and how we handle the difference between those two platforms. For example, If we are going to use `-Wl,--gc-sections` on Linux, then should we put procedures in different sections on macOS?

**Comment by CaviarChen** [comment]
@pschachte In our last meeting, we decided to put each function in a separate section for `-Wl,--gc-sections` on Linux and keep the same behavior on macOS. However, I just realized that the section specifier of ELF and mach-o are different.
In ELF, it's just a section name and in mach-o it contains a segment and section separated by a comma.
For example, the lpvm is stored in `__LPVM.__lpvm` on Linux and `__LPVM,__lpvm` on macOS. And by default all functions are stored in `.text` on Linux and `.TEXT,.text` on macOS.
Currently, for a function foo, we want to put it in `.text.foo` on Linux and I'm not sure where to put it on macOS.
Maybe just keep the old behavior (i.e. don't put it in a separate section) on macOS? Object files are platform-specific stuff and since the section specifier is already different, it shouldn't matter if we don't keep it the same.
